android/app/src/main/java/org/openbot/autopilot/AutopilotFragment.java

```diff
@@ -98,41 +89,12 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
     binding.cameraToggle.setOnClickListener(v -> toggleCamera());
 
     List<String> models =
-        masterList.stream()
-            .filter(f -> f.type.equals(Model.TYPE.AUTOPILOT) && f.pathType != Model.PATH_TYPE.URL)
-            .map(f -> FileUtils.nameWithoutExtension(f.name))
-            .collect(Collectors.toList());
-    modelAdapter = new ArrayAdapter<>(requireContext(), R.layout.spinner_item, models);
-
-    modelAdapter.setDropDownViewResource(android.R.layout.simple_dropdown_item_1line);
-    binding.modelSpinner.setAdapter(modelAdapter);
-    if (!preferencesManager.getAutopilotModel().isEmpty())
-      binding.modelSpinner.setSelection(
-          Math.max(
-              0,
-              modelAdapter.getPosition(
-                  FileUtils.nameWithoutExtension(preferencesManager.getAutopilotModel()))));
+        getModelNames(
+            f -> f.type.equals(Model.TYPE.AUTOPILOT) && f.pathType != Model.PATH_TYPE.URL);
+    initModelSpinner(binding.modelSpinner, models, preferencesManager.getAutopilotModel());
+    initServerSpinner(binding.serverSpinner);
 
     setAnalyserResolution(Enums.Preview.HD.getValue());
```

Comment:
The refactoring is a great idea. But it seems it is the cause for app crash if camera permission is not yet granted.

Reply:
can you provide a stack trace?

Reply:
Did not encounter it anymore now. Maybe it is fixed already.